cli/status.go

```diff
@@ -0,0 +1,214 @@
+package cli
+
+import (
+	"bytes"
+	"fmt"
+	"html/template"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/exercism/cli/user"
+)
+
+// Status represents the results of a CLI self test.
+type Status struct {
+	Censor          bool
+	Version         versionStatus
+	System          systemStatus
+	Configuration   configurationStatus
+	APIReachability apiReachabilityStatus
+	cli             *CLI
+	cfg             user.Config
+}
+
+type versionStatus struct {
+	Current  string
+	Latest   string
+	Status   string
+	Error    error
+	UpToDate bool
+}
+
+type systemStatus struct {
+	OS           string
+	Architecture string
+	Build        string
+}
+
+type configurationStatus struct {
+	Home      string
+	Workspace string
+	File      string
+	Token     string
+	TokenURL  string
+}
+
+type apiReachabilityStatus struct {
+	Services []*apiPing
+}
+
+type apiPing struct {
+	Service string
+	URL     string
+	Status  string
+	Latency time.Duration
+}
+
+func NewStatus(c *CLI, uc user.Config) Status {
+	status := Status{
+		cli: c,
+		cfg: uc,
+	}
+	return status
+}
+
+func (status *Status) Check() (string, error) {
+	status.Version = newVersionStatus(status.cli)
+	status.System = newSystemStatus()
+	status.Configuration = newConfigurationStatus(status)
+	status.APIReachability = newAPIReachabilityStatus()
+
+	return status.compile()
+}
+func (status *Status) compile() (string, error) {
+	t, err := template.New("self-test").Parse(tmplSelfTest)
+	if err != nil {
+		return "", err
+	}
+
+	var bb bytes.Buffer
+	t.Execute(&bb, status)
+	return bb.String(), nil
+}
+
+func newAPIReachabilityStatus() apiReachabilityStatus {
+	ar := apiReachabilityStatus{
+		Services: []*apiPing{
+			{Service: "GitHub", URL: "https://api.github.com"},
+			{Service: "Exercism", URL: "http://exercism.io/api/v1"},
+			{Service: "X-API", URL: "http://x.exercism.io"},
+		},
+	}
+	var wg sync.WaitGroup
+	wg.Add(len(ar.Services))
+	for _, service := range ar.Services {
+		go service.Call(&wg)
+	}
+	wg.Wait()
+	return ar
+}
+
+func newVersionStatus(cli *CLI) versionStatus {
+	vs := versionStatus{
+		Current: cli.Version,
+	}
+	ok, err := cli.IsUpToDate()
+	if err == nil {
+		vs.Latest = cli.LatestRelease.Version()
+	} else {
+		vs.Error = fmt.Errorf("Error: %s", err)
+	}
+	vs.UpToDate = ok
+	return vs
+}
+
+func newSystemStatus() systemStatus {
+	ss := systemStatus{
+		OS:           runtime.GOOS,
+		Architecture: runtime.GOARCH,
+	}
+	if BuildOS != "" && BuildARCH != "" {
```

Comment:
I wanted to avoid the empty `/` in the output, which is confusing even if you are familiar with the codebase :)